jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpChannelOverHTTP2.java

```diff
@@ -261,31 +264,265 @@ public InvocationType getInvocationType()
+        {
+            while (true)
+            {
+                HttpInput.Content c = _content.get();
+                if (c == DEMANDING)
+                {
+                    if (_content.compareAndSet(DEMANDING, content))
+                        return;
+                }
+                else if (c == null)
+                {
+                    if (!content.isSpecial())
+                        throw new AssertionError("Non special content without demand");
+                    if (_content.compareAndSet(null, content))
+                        return;
+                }
+                else if (c.isEof() && content.isEof() && content.isEmpty())
+                {
+                    content.succeeded();
+                    return;
+                }
+                else if (content.getError() != null)
+                {
+                    if (c.getError() != null)
+                    {
+                        if (c.getError() != content.getError())
+                            c.getError().addSuppressed(content.getError());
+                        return;
+                    }
+                    if (_content.compareAndSet(c, content))
+                    {
+                        c.failed(content.getError());
+                        return;
+                    }
+                }
+                else if (c.getError() != null && content.remaining() == 0)
+                {
+                    content.succeeded();
+                    return;
+                }
+                else
+                {
+                    throw new IllegalStateException("Cannot overwrite exiting content " + c + " with " + content);
+                }
+            }
+        }
+
+        public boolean onTimeout(Throwable failure)
+        {
+            while (true)
+            {
+                HttpInput.Content c = _content.get();
+                if (c != DEMANDING)
+                    return false;
+                if (_content.compareAndSet(DEMANDING, new HttpInput.ErrorContent(failure)))
+                    return true;
+            }
+        }
+
+        public void eof()
+        {
+            while (true)
+            {
+                HttpInput.Content c = _content.get();
+                if (c == DEMANDING)
+                {
+                    if (_content.compareAndSet(DEMANDING, EOF))
+                        return;
+                }
+                else if (c == null)
+                {
+                    if (_content.compareAndSet(null, EOF))
+                        return;
+                }
+                else if (c.isEof())
+                {
+                    return;
+                }
+                else if (c.remaining() == 0)
+                {
+                    if (_content.compareAndSet(c, EOF))
+                        return;
+                }
+                else
+                {
+                    HttpInput.Content content = new HttpInput.Content(c.getByteBuffer())
```

Comment:
I think you're getting confused because the `HttpInput.Content` contract isn't documented, so I'll add the necessary javadoc that should cover this question.

But to answer it: special content is special in the sense that it isn't backed by a byte buffer and it just indicates that we reached a special state, either an error or EOF. That, and the `HttpChannel.produceContent()` that specifies that once you reached special content, you'll always get it served.

 @gregw added that non-special content could hint that it is the last content by both being backed by a byte buffer and have the EOF flag set. But it's just a hint that currently only is implemented in H2 and that isn't relied upon anywhere. Everywhere out of the H2 channel, the code only expects EOF that is a special content.

Reply:
Currently EOF can arrive in H2 in two ways: with a data frame that has content; with an empty data frame.

With the former, a Frame with content and EOF will be created and then added, then eof() is called, which will be a noop in the `c.isEof()` branch above.

With the later, the demand should never be done until the frame with content is consumed, so the EOF will only arrive when their is no data.   I think onData will not be called, and then eof() will be called and do the simple `c == null` branch above.

So strictly speaking, this wrapping here of a non EOF content with one that is EOF should never EVER be used.  Perhaps we can just do a ISE or AE here... but then the eof hint on content becomes more than a hint

Reply:
You're right, this branch should never be taken so it can be made to throw AE instead.

Regarding the eof hint, it's not a hint within the context of this implementation, not in protocol-generic code so I think it's fine.